tvm_utility/test/yolo_v2_tiny/main.cpp

```diff
@@ -0,0 +1,283 @@
+          auto w = anchor_w * exp(box_w) / l_w;
+          auto h = anchor_h * exp(box_h) / l_h;
+
+          // Decode the confidence of detection in this anchor box
+          auto p_0 = sigmoid(box_p);
+
+          // find maximum probability of all classes
+          float max_p = 0.0f;
+          int max_ind = -1;
+          for (int i_class = 0; i_class < n_classes; i_class++) {
+            auto class_p = get_output_data(i, j, anchor_k, 5 + i_class);
+            if (max_p < class_p) {
+              max_p = class_p;
+              max_ind = i_class;
+            }
+          }
+
+          // decode and copy class probabilities
+          std::vector<float> class_probabilities{};
+          float p_total = 0;
+          for (size_t i_class = 0; i_class < n_classes; i_class++) {
+            auto class_p = get_output_data(i, j, anchor_k, 5 + i_class);
+            class_probabilities.push_back(std::exp(class_p - max_p));
+            p_total += class_probabilities[i_class];
+          }
+
+          // Find the most likely score
+          auto max_score = class_probabilities[max_ind] * p_0 / p_total;
+
+          // draw all detections with high scores
+          if (max_score > 0.3) {
+            scores_above_threshold.push_back(max_score);
+          }
+        }
+      }
+    }
+
+    return scores_above_threshold;
+  }
+
+private:
+  int64_t network_output_width;
+  int64_t network_output_height;
+  int64_t network_output_depth;
+  std::vector<std::string> labels{};
+  std::vector<std::pair<float, float>> anchors{};
+};
+
+// bring config into scope
+namespace model_config = model_zoo::perception::camera_obstacle_detection::
+    yolo_v2_tiny::tensorflow_fp32_coco;
+using tvm_utility::pipeline::InferenceEngineTVM;
+using tvm_utility::pipeline::Pipeline;
+
+TEST(PipelineExamples, SimplePipeline) {
+  // instantiate the pipeline
+  Pipeline<PreProcessorYoloV2Tiny, InferenceEngineTVM, PostProcessorYoloV2Tiny>
+      pipeline(PreProcessorYoloV2Tiny{model_config::config},
+               InferenceEngineTVM{model_config::config},
+               PostProcessorYoloV2Tiny{model_config::config});
+
+  // push data input the pipeline and get the output
+  sensor_msgs::PointCloud2 msg{};
+  auto output = pipeline.schedule(msg);
+
+  // define reference vector containing expected values
+  std::vector<float> expected_output{0.360896, 0.763201, 0.402838, 0.302775,
+                                     0.612508, 0.583656, 0.422576, 0.452684,
+                                     0.372608, 0.806072, 0.572778};
+
+  // test: check if the generated output is equal to the reference
+  EXPECT_EQ(expected_output.size(), output.size()) << "Unexpected output size";
+  for (auto i = 0; i < output.size(); ++i) {
+    EXPECT_NEAR(expected_output[i], output[i], 0.0001) << "at index: " << i;
+  }
+}
+
+int main(int argc, char **argv) {
+  ::testing::InitGoogleTest(&argc, argv);
+  return RUN_ALL_TESTS();
+}
```

Comment:
The `main` function isn't necessary because `catkin_add_gtest` automatically links against a GTest library target that creates a `main.cpp` file.

Reply:
That doesn't seem to be the case, http://wiki.ros.org/gtest shows that a `main()` function is used with `catkin_add_gtest`. Do you have more information on that?

Reply:
@ambroise-arm This took a little while to find for `catkin` (it's much more obvious in `ament`) but here's the documentation on linking against the `gtest_main` library which provides a default entry point: https://github.com/google/googletest/blob/master/googletest/docs/primer.md#writing-the-main-function

When you call `catkin_add_gtest(<target>)`, the following happens in https://github.com/ros/catkin/blob/noetic-devel/cmake/test/gtest.cmake:
- `catkin_add_gtest` calls `_catkin_add_google_test`
- `_catkin_add_google_test` calls `_catkin_add_executable_with_google_test`
- `_catkin_add_executable_with_google_test` calls `target_link_libraries(<your_target> ${GTEST_LIBRARIES})` [here](https://github.com/ros/catkin/blob/a9672d78ec483c3a991a051c365da329fceaa9f2/cmake/test/gtest.cmake#L186)
- The `CMakeLists.txt` file from `catkin` says that `${GTEST_LIBRARIES}` is defined by the `cmake-extras` package's `GMockConfig.cmake` file, which I found [here](https://github.com/ubports/cmake-extras/blob/master/src/GMock/GMockConfig.cmake) which just defines it to be `gtest`

Unfortunately, it looks like you were correct. If you wanted to simplify this code, you could link the target against `${GTEST_MAIN_LIBRARIES}` and remove this `main` function but this is fine also.